kotti/rest.py

```diff
@@ -0,0 +1,317 @@
+        response.body = render('kotti_jsonp', new_item, self.request)
+        return response
+
+    @view_config(request_method='DELETE', permission='delete')
+    def delete(self):
+        # data = self.request.json_body['data']
+
+        parent = self.context.__parent__
+        del parent[self.context.__name__]
+        return HTTPNoContent()
+
+
+def get_schema(obj, request, name=u'default'):
+    factory_name = _schema_factory_name(context=obj, name=name)
+    schema_factory = request.registry.getUtility(ISchemaFactory,
+                                                 name=factory_name)
+    return schema_factory(obj, request)
+
+
+def get_content_factory(request, name):
+    return request.registry.getUtility(IContentFactory, name=name)
+
+
+# def filter_schema(schema, allowed_fields):
+#     """ Filters a schema to include only allowed fields
+#     """
+#
+#     cloned = schema.__class__(self.typ)
+#     cloned.__dict__.update(schema.__dict__)
+#     cloned.children = [node.clone() for node in self.children
+#                        if node.name in allowed_fields]
+#     return cloned
+#
+
+class MetadataSchema(colander.MappingSchema):
+    """ Schema that exposes some metadata information about a content
+    """
+
+    modification_date = colander.SchemaNode(
+        colander.Date(),
+        title=_(u'Modification Date'),
+    )
+
+    creation_date = colander.SchemaNode(
+        colander.Date(),
+        title=_(u'Modification Date'),
+    )
+
+    state = colander.SchemaNode(
+        colander.String(),
+        title=_(u'State'),
+    )
+    state = colander.SchemaNode(
+        colander.String(),
+        title=_(u'State'),
+    )
+
+    default_view = colander.SchemaNode(
+        colander.String(),
+        title=_(u'Default view'),
+    )
+
+    in_navigation = colander.SchemaNode(
+        colander.String(),
+        title=_(u'In navigation'),
+    )
+
+
+def serialize(obj, request, name=u'default'):
+    """ Serialize a Kotti content item.
+
+    The response JSON conforms with JSONAPI standard.
+
+    TODO: implement JSONAPI filtering and pagination.
+    """
+    data = get_schema(obj, request, name).serialize(obj.__dict__)
+
+    res = {}
+    res['type'] = obj.type_info.name
+    res['id'] = obj.__name__
+    res['attributes'] = data
+    res['links'] = {
+        'self': request.resource_url(obj),
+        'children': [request.resource_url(child)
+                     for child in obj.children_with_permission(request)]
+    }
+    meta = MetadataSchema().serialize(obj.__dict__)
+
+    return dict(data=res, meta=meta)
+
+
+def _encoder(basedefault):
+    """ A JSONEncoder that can encode some basic odd objects.
+
+    For most objects it will execute the basedefault function, which uses
+    adapter lookup mechanism to achieve the encoding, but for some basic
+    objects, such as datetime and colander.null we solve it here.
+    """
+
+    class Encoder(json.JSONEncoder):
+
+        def default(self, obj):
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [An attribute defined in json.encoder line 162 hides this method](https://www.codacy.com/app/disko/Kotti/file/994475488/issues/source?bid=2089747&fileBranchId=2861988#l290)